custom/icds_reports/utils/aggregation_helpers/distributed/agg_dashboard_activity.py

```diff
@@ -0,0 +1,275 @@
+
+        return """
+        INSERT INTO "{tablename}" (
+            username, state_id, district_id,block_id,
+            user_level,date
+        )
+        VALUES {param_keys}
+        """.format(
+            tablename=self.tablename,
+            param_keys=','.join(param_keys)
+        ), parameters
+
+    def rollover_previous_data(self):
+        query_param = {'date': self.date,
+                       'last_agg_date': self.last_agg_date}
+
+        return """
+        UPDATE "{tablename}" user_activity SET
+            location_launched = ut.location_launched,
+            last_activity = ut.last_activity
+        FROM  (
+        SELECT
+            username,
+            location_launched,
+            last_activity
+            FROM "{parent_tablename}" WHERE date = %(last_agg_date)s
+            )ut
+        WHERE user_activity.username = ut.username
+        """.format(
+            tablename=self.tablename,
+            parent_tablename=self.aggregate_parent_table
+        ), query_param
+
+    def update_queries(self):
+        last_time_to_consider = self.date
+        latest_month = (last_time_to_consider - timedelta(days=1)).replace(day=1)
+
+        yield """
+        UPDATE "{tablename}" user_activity
+            SET location_launched = CASE WHEN num_launched_blocks>0 THEN TRUE ELSE FALSE END
+        FROM (
+            SELECT
+                block_id, district_id, state_id, num_launched_blocks,aggregation_level
+            FROM agg_awc where month=%(latest_month)s and aggregation_level<=3
+            ) ut
+        WHERE (
+            user_activity.user_level= ut.aggregation_level AND
+            user_activity.state_id = ut.state_id AND
+            user_activity.district_id = ut.district_id AND
+            user_activity.block_id = ut.block_id AND
+            user_activity.location_launched is not TRUE
+        )
+        """.format(
+            tablename=self.tablename
+        ), {
+            'latest_month': latest_month
+        }
+
+        # This is query I prepare which could do what the last two queries are doing but
+        # the cost of this single query coming out to be very high
+        # yield """
+        # UPDATE {tablename} user_activity
+        # SET
+        #     last_activity
+        # from (
+        # SELECT audit.username, max(audit.time_of_use) from {tablename} user_activity left join
+        # icds_audit_entry_record audit
+        # on user_activity.username = audit.username
+        # group by username where (user_activity.last_activity is not null AND
+        #                          time_of_use>user_activity.last_activity
+        #                           ) or
+        #                         ( user_activity.last_activity is null)
+        # )ut
+        # where user_activity.username = ut.username
+        # """.format(tablename=self.tablename)
+
+        yield """
+        UPDATE "{tablename}" user_activity
```

Comment:
Cost:
```
Update on temp_rohit_activity user_activity  (cost=824260.67..824371.52 rows=34254 width=113)
   ->  Hash Join  (cost=824260.67..824371.52 rows=34254 width=113)
         Hash Cond: ((user_activity.username)::text = (ut.username)::text)
         ->  Seq Scan on temp_rohit_activity user_activity  (cost=0.00..22.00 rows=1200 width=38)
         ->  Hash  (cost=824189.30..824189.30 rows=5709 width=110)
               ->  Subquery Scan on ut  (cost=824075.12..824189.30 rows=5709 width=110)
                     ->  HashAggregate  (cost=824075.12..824132.21 rows=5709 width=43)
                           Group Key: icds_audit_entry_record.username
                           ->  Nested Loop  (cost=25.54..699734.60 rows=24868104 width=43)
                                 ->  HashAggregate  (cost=24.98..26.98 rows=200 width=32)
                                       Group Key: (temp_rohit_activity.username)::text
                                       ->  Seq Scan on temp_rohit_activity  (cost=0.00..22.00 rows=1194 width=32)
                                             Filter: (last_activity IS NOT NULL)
                                 ->  Append  (cost=0.56..3231.53 rows=26701 width=43)
                                       ->  Index Scan using icds_audit_entry_record_username_ed335fe7 on icds_audit_entry_record  (cost=0.56..764.35 rows=2470 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: (time_of_use <= '2019-12-09 00:00:00+00'::timestamp with time zone)
                                       ->  Index Scan using icds_audit_entry_record_y2019m06_username_idx on icds_audit_entry_record_y2019m06  (cost=0.43..104.56 rows=796 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: (time_of_use <= '2019-12-09 00:00:00+00'::timestamp with time zone)
                                       ->  Index Scan using icds_audit_entry_record_y2018m11_username_idx on icds_audit_entry_record_y2018m11  (cost=0.55..132.44 rows=2646 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: (time_of_use <= '2019-12-09 00:00:00+00'::timestamp with time zone)
                                       ->  Index Scan using icds_audit_entry_record_y2018m10_username_idx on icds_audit_entry_record_y2018m10  (cost=0.42..12.33 rows=246 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: (time_of_use <= '2019-12-09 00:00:00+00'::timestamp with time zone)
                                       ->  Index Scan using icds_audit_entry_record_y2018m05_username_idx1 on icds_audit_entry_record_y2018m05  (cost=0.42..11.92 rows=244 width=44)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: (time_of_use <= '2019-12-09 00:00:00+00'::timestamp with time zone)
                                       ->  Index Scan using icds_audit_entry_record_y2018m06_username_idx on icds_audit_entry_record_y2018m06  (cost=0.42..43.79 rows=735 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: (time_of_use <= '2019-12-09 00:00:00+00'::timestamp with time zone)
                                       ->  Index Scan using icds_audit_entry_record_y2018m07_username_idx on icds_audit_entry_record_y2018m07  (cost=0.42..66.14 rows=1041 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: (time_of_use <= '2019-12-09 00:00:00+00'::timestamp with time zone)
                                       ->  Index Scan using icds_audit_entry_record_y2018m08_username_idx on icds_audit_entry_record_y2018m08  (cost=0.56..273.96 rows=5146 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: (time_of_use <= '2019-12-09 00:00:00+00'::timestamp with time zone)
                                       ->  Index Scan using icds_audit_entry_record_y2019m03_username_idx on icds_audit_entry_record_y2019m03  (cost=0.43..65.16 rows=883 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: (time_of_use <= '2019-12-09 00:00:00+00'::timestamp with time zone)
                                       ->  Index Scan using icds_audit_entry_record_y2019m04_username_idx on icds_audit_entry_record_y2019m04  (cost=0.43..103.29 rows=1440 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: (time_of_use <= '2019-12-09 00:00:00+00'::timestamp with time zone)
                                       ->  Index Scan using icds_audit_entry_record_y2019m01_username_idx on icds_audit_entry_record_y2019m01  (cost=0.43..107.46 rows=1958 width=42)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: (time_of_use <= '2019-12-09 00:00:00+00'::timestamp with time zone)
                                       ->  Index Scan using icds_audit_entry_record_y2018m12_username_idx on icds_audit_entry_record_y2018m12  (cost=0.43..77.29 rows=1449 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: (time_of_use <= '2019-12-09 00:00:00+00'::timestamp with time zone)
                                       ->  Index Scan using icds_audit_entry_record_y2019m02_username_idx on icds_audit_entry_record_y2019m02  (cost=0.56..149.02 rows=2065 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: (time_of_use <= '2019-12-09 00:00:00+00'::timestamp with time zone)
                                       ->  Index Scan using icds_audit_entry_record_y2018m09_username_idx on icds_audit_entry_record_y2018m09  (cost=0.42..26.05 rows=528 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: (time_of_use <= '2019-12-09 00:00:00+00'::timestamp with time zone)
                                       ->  Index Scan using icds_audit_entry_record_y2019m05_username_idx on icds_audit_entry_record_y2019m05  (cost=0.56..166.51 rows=1599 width=43)
                                             Index Cond: ((username)::text = (temp_rohit_activity.username)::text)
                                             Filter: (time_of_use <= '2019-12-09 00:00:00+00'::timestamp with time zone)
```

Reply:
I have tested this will a dummy report of only 3 users.
I think the cost of this is going to increase as the report will be of 15K users. Looking that I think your suggestion of one query is better. But will let you comment if you still feel so.

Reply:
This query will run every time not just the first time. That is why I think your suggestion is better. But that single query is showing extremely high cost `4316146.33` which I don't know how DB will perform. So request your input.

Reply:
Had discussed this with Emord, and it is fine as this will only for first time. after that cost will be very much less.